App: Allow the user to pass arbitrary scalac options

pingtimeout opened the pull request:
Do not merge yet: not tested

With the DSE java-driver, we need to pass `-Ybreak-cycles` to the scala compiler. Instead of hardcoding that option in `ZincCompiler`, this PR introduces a new command line parameter: `-eso` or `--extra-scalac-options` that allows arbitrary options to be passed.

A new environment variable has been added in gatling.sh: EXTRA_SCALAC_OPTIONS. If it is set, then its content is used as the value for `-eso`. I added it because I think compiler arguments should not be passed in `$@` and end up in the real Gatling invocation.

Not tested yet: for some reason, I get the following behavior on `master`.  I reproduced the problem without the patch, which makes me think there is something wrong on master.  Does anyone have the same issue?

```bash
$ ./bin/gatling.sh
GATLING_HOME is set to /tmp/gatling-charts-highcharts-bundle-3.0.0-SNAPSHOT
There is no simulation script. Please check that your scripts are in user-files/simulations
```

pingtimeout commented:
@slandelle by any chance, could you confirm that there is an issue one master that prevents the bundle from detecting simulations?

slandelle commented:
?
Works just fine for me:

```
$ ./gatling.sh
GATLING_HOME is set to /Users/slandelle/.m2/repository/io/gatling/highcharts/gatling-charts-highcharts-bundle/3.0.0-SNAPSHOT/gatling-charts-highcharts-bundle-3.0.0-SNAPSHOT
Choose a simulation number:
     [0] computerdatabase.BasicSimulation
     [1] computerdatabase.advanced.AdvancedSimulationStep01
     [2] computerdatabase.advanced.AdvancedSimulationStep02
     [3] computerdatabase.advanced.AdvancedSimulationStep03
     [4] computerdatabase.advanced.AdvancedSimulationStep04
     [5] computerdatabase.advanced.AdvancedSimulationStep05
```

pingtimeout commented:
This is strange... I just rebuild gatling and gatling-highcharts from master (`d1f866d40`) and got this issue.  I made sure to remove all `3.0.0-SNAPSHOT` artifacts I had in my local repos (ivy and m2) to be sure I had a clean state.  I must be missing something...

pingtimeout commented:
Well, my bad, deleting the project and starting again (`git clone`) solved the issue...  Sorry

pingtimeout commented:
Now it is good for review, I tested it and got the following results:

```sh
$ ./bin/gatling.sh
GATLING_HOME is set to /tmp/2018-07-17-gatling-3.0.0-patched/gatling-charts-highcharts-bundle-3.0.0-SNAPSHOT
17:21:27.148 [ERROR] i.g.c.ZincCompiler$ - /tmp/2018-07-17-gatling-3.0.0-patched/gatling-charts-highcharts-bundle-3.0.0-SNAPSHOT/user-files/simulations/computerdatabase/BasicSimulation.scala:11:36: illegal cyclic reference involving class Cluster
  val cluster = DseCluster.builder().addContactPoint("127.0.0.1").build()
                                   ^
17:21:27.170 [ERROR] i.g.c.ZincCompiler$ - /tmp/2018-07-17-gatling-3.0.0-patched/gatling-charts-highcharts-bundle-3.0.0-SNAPSHOT/user-files/simulations/computerdatabase/BasicSimulation.scala:13:54: illegal cyclic reference involving class Cluster
  val statement: PreparedStatement = session.prepare("SELECT now() FROM system.local;")
                                                     ^
17:21:30.675 [ERROR] i.g.c.ZincCompiler$ - two errors found
17:21:30.677 [ERROR] i.g.c.ZincCompiler$ - Compilation crashed
...

$ EXTRA_SCALAC_OPTIONS="-Ybreak-cycles" ./bin/gatling.sh
GATLING_HOME is set to /tmp/2018-07-17-gatling-3.0.0-patched/gatling-charts-highcharts-bundle-3.0.0-SNAPSHOT
17:23:40.016 [WARN ] i.g.c.ZincCompiler$ - Breaking cycle in base class computation of class Cluster in com.datastax.driver.core (List(class Cluster, trait Closeable, trait AutoCloseable, class Object, class Any))
17:23:49.526 [WARN ] i.g.c.ZincCompiler$ - one warning found
Choose a simulation number:
     [0] computerdatabase.BasicSimulation
     [1] computerdatabase.advanced.AdvancedSimulationStep01
     [2] computerdatabase.advanced.AdvancedSimulationStep02
     [3] computerdatabase.advanced.AdvancedSimulationStep03
     [4] computerdatabase.advanced.AdvancedSimulationStep04
     [5] computerdatabase.advanced.AdvancedSimulationStep05
^C
```

I could not use the gatling-dse-plugin to test this simulation, as it still uses classes from the 2.x Gatling branch (`Extender` and other checker related classes).  To test this, I added the [DSE java-driver](http://search.maven.org/#artifactdetails%7Ccom.datastax.dse%7Cdse-java-driver-core%7C1.6.8%7Cjar) in the `lib/` folder and changed the BasicSimulation so that it opens a connection:

```scala
import com.datastax.driver.core.PreparedStatement
import com.datastax.driver.dse.{DseCluster, DseSession}

class BasicSimulation extends Simulation {

  val cluster = DseCluster.builder().addContactPoint("127.0.0.1").build()
  val session: DseSession = cluster.connect()
  val statement: PreparedStatement = session.prepare("SELECT now() FROM system.local;")

  val httpConf = http
  ...
```

Note that a DSE cluster is not needed to verify the compilation and the effect of `-Ybreak-cycles`.

slandelle commented:
LGTM

Could you please:
* squash your commits into a single one
* sign our CLA: https://docs.google.com/forms/d/1hbpxVqJ5hIYYJuOjOyOzub9fJFGFGvvfnB5w5LCWdUw

pingtimeout commented:
Done²

slandelle commented:
cherry-picked, thanks!
